sdk/trace/trace_test.go

```diff
@@ -1594,6 +1594,90 @@ func TestAddLinksWithMoreAttributesThanLimit(t *testing.T) {
 	}
 }
 
+type frozenClock struct {
+	now time.Time
+}
+type frozenStopwatch struct {
+	started time.Time
+}
+
+func (f frozenStopwatch) Started() time.Time {
+	return f.started
+}
+func (f frozenStopwatch) Stop() time.Duration {
+	return 0
+}
+
+// newFrozenClock returns a clock which stops at time t
+func newFrozenClock(t time.Time) frozenClock {
+	return frozenClock{
+		now: t,
+	}
+}
+
+func (c frozenClock) Stopwatch() Stopwatch {
+	return frozenStopwatch{
+		started: c.now,
+	}
+}
+
+type backwardClock struct{}
+type backwardStopwatch struct {
+	started time.Time
+}
+
+func (b backwardStopwatch) Started() time.Time {
+	return b.started
+}
+
+func (b backwardStopwatch) Stop() time.Duration {
+	return -time.Since(b.started)
+}
+
+func (b backwardClock) Stopwatch() Stopwatch {
+	return backwardStopwatch{
+		started: time.Now(),
+	}
+}
+
+func TestCustomClock(t *testing.T) {
+	te := NewTestExporter()
+	tp := NewTracerProvider(WithSyncer(te))
+	tracer := tp.Tracer("custom-clock")
+	now := time.Now()
+	time.Sleep(time.Microsecond * 3)
+	_, span := tracer.Start(context.Background(), "test-original-clock")
+	time.Sleep(time.Microsecond * 3)
+	span.End()
+	got := te.Spans()[0]
+	assert.True(t, now.Before(got.StartTime()))
+	assert.True(t, got.StartTime().Before(got.EndTime()))
+
+	te.Reset()
+
+	tp = NewTracerProvider(WithSyncer(te), WithClock(newFrozenClock(now)))
+	tracer = tp.Tracer("custom-clock")
+	time.Sleep(time.Microsecond * 3)
+	_, span = tracer.Start(context.Background(), "test-frozen-clock")
+	time.Sleep(time.Microsecond * 3)
+	span.End()
+	got = te.Spans()[0]
+	assert.Equal(t, now, got.StartTime())
+	assert.Equal(t, now, got.EndTime())
+
+	te.Reset()
+	tp = NewTracerProvider(WithSyncer(te), WithClock(backwardClock{}))
+	tracer = tp.Tracer("custom-clock")
+	now = time.Now()
+	time.Sleep(time.Microsecond * 3)
+	_, span = tracer.Start(context.Background(), "test-backward-clock")
+	time.Sleep(time.Microsecond * 3)
+	span.End()
+	got = te.Spans()[0]
+	assert.True(t, now.Before(got.StartTime()))
+	assert.True(t, got.StartTime().After(got.EndTime()))
+}
```

Comment:
If I remember correctly, on windows the clock seems to have a lower precision which will result in the same `now` and `startTime` regardless of the custom clock used (which makes this test a little bit of meaningless on windows). So that's why `sleep` was introduced here. How about reduce it to 1 microsecond each call?
And I think keep only the `newFrozenClock` is OK given that they actually all test the same logic.

Reply:
> How about reduce it to 1 microsecond each call?

I think we just need a single `time.Sleep(2 * time.Microsecond)` just after `_, span = tracer.Start(context.Background(), "test-frozen-clock")`

> And I think keep only the newFrozenClock is OK given that they actually all test the same logic.

👍

Reply:
updated 😄